File: data_structures/trie/cpp/trie.cpp
#include "trie.h"

Trie::Trie() : root_(new Node{}) {}

Trie::~Trie() { destroy(root_); }

void Trie::destroy(Node* n) {
    if (!n) return;
    for (auto* c : n->children) destroy(c);
    delete n;
}

void Trie::insert(const std::string& word) {
    Node* cur = root_;
    for (char ch : word) {
        int idx = ch - 'a';
        if (!cur->children[idx]) cur->children[idx] = new Node{};
        cur = cur->children[idx];
        ++cur->count;
    }
    if (!cur->is_end) { cur->is_end = true; ++size_; }
}

bool Trie::search(const std::string& word) const {
    Node* cur = root_;
    for (char ch : word) {
        int idx = ch - 'a';
        if (!cur->children[idx]) return false;
        cur = cur->children[idx];
    }
    return cur->is_end;
}

bool Trie::starts_with(const std::string& prefix) const {
    Node* cur = root_;
    for (char ch : prefix) {
        int idx = ch - 'a';
        if (!cur->children[idx]) return false;
        cur = cur->children[idx];
    }
    return true;
}

bool Trie::remove(Node* n, const std::string& word, int depth, bool& removed) {
    if (!n) return false;
    if (depth == static_cast<int>(word.size())) {
        if (!n->is_end) return false;
        n->is_end = false;
        removed = true;
        return n->count == 0;
    }
    int idx = word[depth] - 'a';
    if (!n->children[idx]) return false;
    bool should_delete = remove(n->children[idx], word, depth + 1, removed);
    if (removed) --n->children[idx]->count;
    if (should_delete) {
        delete n->children[idx];
        n->children[idx] = nullptr;
    }
    return !n->is_end && n->count == 0;
}

bool Trie::remove(const std::string& word) {
    bool removed = false;
    remove(root_, word, 0, removed);
    if (removed) --size_;
    return removed;
}

int Trie::size() const { return size_; }

File: data_structures/trie/python/trie.py
"""トライ木（前置木）

操作:
  insert(word)        : 文字列を挿入  O(L)
  search(word)        : 文字列を検索  O(L)
  starts_with(prefix) : プレフィックス検索  O(L)
  remove(word)        : 文字列を削除  O(L)
  size()              : 挿入済み単語数  O(1)

小文字英字 a-z のみ対応。
"""

from typing import Dict, Optional


class _Node:
    __slots__ = ("children", "is_end", "count")

    def __init__(self) -> None:
        self.children: Dict[str, "_Node"] = {}
        self.is_end = False
        self.count = 0


class Trie:
    def __init__(self) -> None:
        self._root = _Node()
        self._size = 0

    def insert(self, word: str) -> None:
        cur = self._root
        for ch in word:
            if ch not in cur.children:
                cur.children[ch] = _Node()
            cur = cur.children[ch]
            cur.count += 1
        if not cur.is_end:
            cur.is_end = True
            self._size += 1

    def search(self, word: str) -> bool:
        cur = self._root
        for ch in word:
            if ch not in cur.children:
                return False
            cur = cur.children[ch]
        return cur.is_end

    def starts_with(self, prefix: str) -> bool:
        cur = self._root
        for ch in prefix:
            if ch not in cur.children:
                return False
            cur = cur.children[ch]
        return True

    def remove(self, word: str) -> bool:
        removed = [False]

        def _remove(node: Optional[_Node], depth: int) -> bool:
            if node is None:
                return False
            if depth == len(word):
                if not node.is_end:
                    return False
                node.is_end = False
                removed[0] = True
                return node.count == 0
            ch = word[depth]
            if ch not in node.children:
                return False
            should_delete = _remove(node.children[ch], depth + 1)
            if removed[0]:
                node.children[ch].count -= 1
            if should_delete:
                del node.children[ch]
            return not node.is_end and node.count == 0

        _remove(self._root, 0)
        if removed[0]:
            self._size -= 1
        return removed[0]

    def size(self) -> int:
        return self._size
